File: DemoObjC/ViewController.m
#import "ViewController.h"
#import "../MGKMenuWithFilter.h"
#import <Quartz/Quartz.h>

@interface ViewController ()

@property (strong) IBOutlet NSButton *pushMeButton;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view.
}


- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

- (IBAction)showPopUpMenu:(id)sender
{
    NSURL *folder = [NSURL fileURLWithPath:@"/Library/Desktop Pictures"];

    NSArray *all_files = [NSFileManager.defaultManager contentsOfDirectoryAtURL:folder
                                                     includingPropertiesForKeys:nil
                                                                        options:0
                                                                          error:nil];
    
    NSString *jpg_predicate_string = @"self.absoluteString ENDSWITH '.jpg'";
    NSPredicate *jpg_predicate = [NSPredicate predicateWithFormat:jpg_predicate_string];
    NSArray *jpg_files = [all_files filteredArrayUsingPredicate:jpg_predicate];
    
    MGKMenuWithFilter *menu = [[MGKMenuWithFilter alloc] initWithTitle:@"Pictures"];
    
    for( NSURL *picture in jpg_files ) {
        NSMenuItem *item = [[NSMenuItem alloc] init];
        item.title = picture.lastPathComponent;
        item.action = @selector(showPicture:);
        item.target = self;
        
        NSSize size = NSMakeSize(32, 32);
        CGImageRef preview = QLThumbnailImageCreate(0, (__bridge CFURLRef)picture, size, 0);
        if( preview ) {
            item.image = [[NSImage alloc] initWithCGImage:preview size:size];
            CGImageRelease(preview);
        }

        item.representedObject = picture;
        [menu addItem:item];
    }
    
    [menu popUpMenuPositioningItem:nil
                         atLocation:NSMakePoint(0, self.pushMeButton.bounds.size.height)
                             inView:self.pushMeButton];
}

- (IBAction)showPicture:(id)sender
{
    [NSWorkspace.sharedWorkspace openURL:[sender representedObject]];
}

@end

File: DemoSwift/ViewController.swift
import Cocoa
import Quartz

class ViewController: NSViewController {

    @IBOutlet var pushMeButton: NSButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }

    @IBAction func showPopUpMenu(_ sender: Any) {
        let folder = NSURL.fileURL(withPath: "/Library/Desktop Pictures")

        let all_files = try! FileManager.default.contentsOfDirectory(at:folder,
                                                               includingPropertiesForKeys:nil,
                                                               options:.skipsSubdirectoryDescendants)
        
        let jpg_files = all_files.filter() { $0.absoluteString.hasSuffix(".jpg") }
        

        let menu = MGKMenuWithFilter.init(title: "Pictures")!
        
        for picture in jpg_files {
            let item = NSMenuItem()
            
            item.title = picture.lastPathComponent
            
            item.action = #selector(showPicture(_:))
            item.target = self

            let size = NSMakeSize(32, 32)
            let preview = QLThumbnailImageCreate(nil, picture as CFURL!, size, nil as CFDictionary!)
            if preview != nil  {
                item.image = NSImage.init(cgImage:preview!.takeUnretainedValue(), size: size)
                preview!.release()
            }
            
            item.representedObject = picture

            menu.addItem(item)
        }
    
        menu.popUp(positioning: nil,
                   at: NSMakePoint(0, pushMeButton.bounds.size.height),
                   in: pushMeButton)
    
    }
    
    @IBAction func showPicture(_ sender: Any) {
        NSWorkspace.shared().open((sender as! NSMenuItem).representedObject as! URL)
    }

}

